### ISSUE TITLE
attempt to delete creates broken resource

### ISSUE THREAD

**Issue by dtarb** [hydroshare/hydroshare#3713]
I  attempted to delete the following resource created in August
https://www.hydroshare.org/resource/2c463f6a285147598d8992ffde93c1f2/

It is a test resource that appeared to be empty.  I got the error "sorry an error occurred".  I believe these resources were created when testing and debugging the problem that deleting an aggregation then put the resource in a state where no new aggregations could be created #3543 (though recollection may be flaky)

Now when I return to the landing page I see

![image](https://user-images.githubusercontent.com/3456833/70376540-1ea66000-18c7-11ea-9f0f-a44e0fbc2aaf.png)

Note that authors and owners are empty and the content spinner goes on indefinitely

Clicking on manage access results in
![image](https://user-images.githubusercontent.com/3456833/70376571-4e556800-18c7-11ea-949e-50ad63548f13.png)

and manage access is non functional, so I am not able to share this with the developer group or any developers to enable you to look at it.  You may need to look at it as admin.

I also tried to delete the resource

https://www.hydroshare.org/resource/9dd9057fd4fa4507a8f6dd355ad4eba1/

I also received "sorry an error occurred"

Before trying to delete, this resource had a text content file

After trying to delete, the contents area indicates no files, but it still has a size.

![image](https://user-images.githubusercontent.com/3456833/70376601-b2782c00-18c7-11ea-8d61-554349653e80.png)

I then tried to delete again and got into the state below, with spinner again going indefinitely and manage access disfunctional.  I checked manage access before trying to delete and this resource was shared with the developers group.
![image](https://user-images.githubusercontent.com/3456833/70376691-f3bd0b80-18c8-11ea-8b2c-c25359b83049.png)

**Comment by martinseul** [comment]
i did look into it as well and can confirm the behavior for my account, deleting "removes" author and abstract. I did do an additional test I logged in as the admin and then the author and abstract appear and i can correctly delete the resource.

**Comment by sblack-usu** [comment]
Here are the logs, first impression says trouble communicating with iRODS.  I'm looking into the error.
```
[07/Dec/2019 14:42:13] ERROR [django.request:135] Internal Server Error: /hsapi/_internal/2c463f6a285147598d8992ffde93c1f2/delete-resource/
SessionException: (SessionException(...), 'Error processing IRODS request: 3. stderr follows:\n\nremote addresses: 152.54.2.121, 192.168.0.3 ERROR: rmUtil: rm error for /hydroshareZone/home/wwwHydroProxy/2c463f6a285147598d8992ffde93c1f2, status = -105000 status = -105000 SYS_RESC_IS_DOWN\n')
[07/Dec/2019 14:55:30] ERROR [hs_core.hydroshare.resource:57] no quota holder is found for resource2c463f6a285147598d8992ffde93c1f2
[07/Dec/2019 14:55:30] ERROR [django.request:135] Internal Server Error: /hsapi/_internal/2c463f6a285147598d8992ffde93c1f2/delete-resource/
SessionException: (SessionException(...), 'Error processing IRODS request: 3. stderr follows:\n\nremote addresses: 152.54.2.121, 192.168.0.3 ERROR: rmUtil: rm error for /hydroshareZone/home/wwwHydroProxy/bags/2c463f6a285147598d8992ffde93c1f2.zip, status = -105000 status = -105000 SYS_RESC_IS_DOWN\n')
```

**Comment by sblack-usu** [comment]
From iRODS documentation: https://docs.irods.org/4.2.2/system_overview/troubleshooting/
>Removing a Data Object, with the force flag, consisting of a single replica, on a resource currently marked down:

>    results in a SYS_RESC_IS_DOWN error
    the Data Object is moved to /[myZone]/trash/orphan/[myUser]#[myZone]/[filename].[10digitrandomsuffix]
    no data movement on disk

>Removing a Data Object, without the force flag, consisting of one replica on a resource marked up, and the other on a resource marked down:

>    results in a SYS_RESC_IS_DOWN error
    catalog is updated, Data Object now in the trash, as expected
    replica on up resource, phypath updated, file moved on disk accordingly
    replica on down resource, phypath untouched, file on disk untouched

>Removing a Data Object, with the force flag, consisting of one replica on a resource marked up, and the other on a resource marked down:

>    results in a SYS_RESC_IS_DOWN error
    catalog is updated, Data Object now at /[myZone]/trash/orphan/[myUser]#[myZone]/[filename].[10digitrandomsuffix]
    replica on up resource is gone
    replica on down resource, phypath untouched, file on disk untouched

**Comment by sblack-usu** [comment]
From the iRODS documentation about this error and some testing, I believe this problem affects resources created before the S3 replica syncing was turned off.  Newly created resources with files, delete fine.  But if I go back to older resources and try to delete one, I am able to reproduce the issue.

I'm going to run some queries and look at the state of resources in production and look into possible hotfixes.  This will be a challenge to create a hotfix for because replicas only exist in www... I'll have to talk with @hyi and @cbcunc, but we probably can just try catch the iRODS delete and handle this error. 

 @dtarb - based off this, I believe you should be fine as long as you don't try to delete resources created before the S3 replica syncing was turned off.  Are you able to navigate to my-resources?  I can finish deleting that resource for you if it is causing problems.

**Comment by sblack-usu** [comment]
Here is the code for deleting.  Looks like metadata should be intact but aggregations will be deleted (looks to be true with this resource).  I would expect my resources to work fine for @dtarb.
```
    def delete(self, using=None):
        """Delete resource along with all of its metadata and data bag."""
        from .hydroshare import hs_bagit
        for fl in self.files.all():
            if fl.logical_file is not None:
                # delete of metadata file deletes the logical file (one-to-one relation)
                # so no need for fl.logical_file.delete() and deleting of metadata file
                # object deletes (cascade delete) all the contained GenericRelated metadata
                # elements
                fl.logical_file.metadata.delete()
            # COUCH: delete of file objects now cascades.
            fl.delete()
        hs_bagit.delete_files_and_bag(self)
        # TODO: Pabitra - delete_all_elements() may not be needed in Django 1.8 and later
        self.metadata.delete_all_elements()
        self.metadata.delete()
        super(AbstractResource, self).delete()
```
The error occurred while in `hs_bagit.delete_files_and_bag(self)`

**Comment by sblack-usu** [comment]
didn't mean to close, reopened.

**Comment by dtarb** [comment]
@sblack-usu I can confirm that "my resources" and my "home" pages both work and list these resources.  They even list the authors, even though the resource landing pages no longer list authors.

**Comment by dtarb** [comment]
@sblack-usu if you are easily and safely able to finish deleting those resources please do so.  That will leave me with a cleaner interface for any demos I do.  However if you want to keep these as illustrations, or are worried about the safety of this in terms of other possible consequences, they can be left.

**Comment by alvacouch** [comment]
@dtarb @sblack-usu @martinseul I have management commands that do recovery for improperly deleted resources and ensure that they are really deleted. 

However, there is a general fix for the behavior we see, which is to trap the iRODs error and ensure that the resource is deleted from Django even if it is not deleted from iRODs. This will leave iRODS corruption behind, but that can be removed later and will be invisible to the user. The only thing the user will see is that the Django record is gone. 

Your thoughts?

**Comment by dtarb** [comment]
I am OK with this approach.

This brings to mind additional considerations.

1. We have in the past discussed having accidental delete protection/recovery options.  If we just delete the record from Django but leave the files in iRODS for a period this may facilitate this.
2. We have also in the past discussed establishing capability for ingesting resources from bags.  If the bag for a deleted resource is complete and we have bag ingestion, this may help with delete recovery.  The delete action would just remove the Django record but leave the resource in iRODS (for whatever period is deemed appropriate).  A recovery request could then be done by ingesting the bag.

These are just my thought.

**Comment by sblack-usu** [comment]
@cbcunc - here is another resource with the problem.  Any resource created before August seems to have this issue (It could be sooner than that).  https://www.hydroshare.org/resource/0aaa134ab4514effb00fed28fa18fdde/

**Comment by cbcunc** [comment]
It has been traced to data objects which have replicas on the hydrodata2Resource iRODS resource. This is the campus replica resource server which had the hard-drive crash. The resource was marked as down two weeks ago. However, deleting a data object which has one good replica and another on a downed resource causes iRODS to place the data object in trash instead of deleting, even with the -f option. Worked with the iRODS team and now a script is running to remove those replicas on the downed resource from the catalog. However, deleting a HS resource is an iRODS I/O operation and should be wrapped in a try/except block.

**Comment by cbcunc** [comment]
Fixed by #3715

**Comment by cbcunc** [comment]
Released as 1.27.3. 72d6630200cf80e4b0036a3ec18c481a0d29e208
